core/src/test/scala/unit/kafka/controller/ControllerIntegrationTest.scala

```diff
@@ -490,6 +490,63 @@ class ControllerIntegrationTest extends ZooKeeperTestHarness {
     }, "failed to get expected partition state after entire isr went offline")
   }
 
+  @Test
+  def testControllerChangeDoesNotPutReplicasOffline(): Unit = {
+    val expectedReplicaAssignment = Map(0  -> List(0, 1))
+    val topic = "test"
+    val partition = 0
+
+    // create brokers
+    val serverConfigs = TestUtils.createBrokerConfigs(3, zkConnect, false).map(KafkaConfig.fromProps)
+    servers = serverConfigs.reverseMap(s => TestUtils.createServer(s))
+    // create the topic
+    TestUtils.createTopic(zkClient, topic, partitionReplicaAssignment = expectedReplicaAssignment, servers = servers)
+    assertTrue(servers.forall(_.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic,partition).get.leader == 0))
+
+    var controllerId = zkClient.getControllerId.get
+    var controller = servers.find(p => p.config.brokerId == controllerId).get.kafkaController
+
+    val resultQueue = new LinkedBlockingQueue[Try[collection.Set[TopicPartition]]]()
+    val controlledShutdownCallback = (controlledShutdownResult: Try[collection.Set[TopicPartition]]) => resultQueue.put(controlledShutdownResult)
+    controller.controlledShutdown(1, servers.find(_.config.brokerId == 1).get.kafkaController.brokerEpoch, controlledShutdownCallback)
+
+    var partitionsRemaining = resultQueue.take().get
+    var activeServers = servers.filter(s => s.config.brokerId != 1)
+    // wait for the update metadata request to trickle to the brokers
+    TestUtils.waitUntilTrue(() =>
+      activeServers.forall(_.dataPlaneRequestProcessor.metadataCache.getPartitionInfo(topic,partition).get.isr.size != 3),
+      "Topic test not created after timeout")
+    assertEquals(0, partitionsRemaining.size)
+
+    servers.find(_.config.brokerId == 1).get.shutdown()
```

Comment:
the real shutdown will trigger the controlled shutdown, so please consider the lines 510 - 519 are still necessary.

Reply:
If we just call `shutdown`, we will not be able to attach the callback and check the partitionsRemaining. Also, shutdown will succeed after configured number of retries. Hence, I think the lines 510-519 are still necessary.

Reply:
I realized that we don't need to issue controlled shutdown for broker 1 and can just call `shutdown`, because what we are really concerned with in this test is the behavior for broker 0 and not broker 1.

Reply:
Yes, that's what I noticed as well.